From ad729f7c83d57ea0843e7a60c8b5605dd4a2b577 Mon Sep 17 00:00:00 2001
From: Bharat Bhushan <bharat.bhushan@nxp.com>
Date: Fri, 1 Mar 2019 13:33:58 +0800
Subject: [PATCH 050/235] vfio/fsl-mc: Improve send mc-command and read
 response

Actually there is no ordering need when reading response
from mc-portal. Similarly when writing the mc-command,
ordering needed before command is submitted.

This patch removes un-necessary barriers, response is read
relaxed and maintain ordering when submit command. This also
fixes compilation issue with newer kernel.

Signed-off-by: Bharat Bhushan <Bharat.Bhushan@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
[Xulin: Original patch taken from NXP LSDK-19.06.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/vfio/fsl-mc/vfio_fsl_mc.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc.c b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
index 1918f342234c..6ae29fac2740 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc.c
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
@@ -336,10 +336,7 @@ static int vfio_fsl_mc_dprc_wait_for_response(void __iomem *ioaddr)
 		u64 header;
 		struct mc_cmd_header *resp_hdr;
 
-		u64 __v = readq_relaxed(ioaddr);
-		__iormb(__v);
-		header = readq(ioaddr);
-		__iormb(__v);
+		header = cpu_to_le64(readq_relaxed(ioaddr));
 
 		resp_hdr = (struct mc_cmd_header *)&header;
 		status = (enum mc_cmd_status)resp_hdr->status;
@@ -359,9 +356,12 @@ static int vfio_fsl_mc_send_command(void __iomem *ioaddr, uint64_t *cmd_data)
 {
 	int i;
 
-	/* Write at command header in the end */
-	for (i = 7; i >= 0; i--)
-		writeq(cmd_data[i], ioaddr + i * sizeof(uint64_t));
+	/* Write at command parameter into portal */
+	for (i = 7; i >= 1; i--)
+		writeq_relaxed(cmd_data[i], ioaddr + i * sizeof(uint64_t));
+
+	/* Write command header in the end */
+	writeq(cmd_data[0], ioaddr);
 
 	/* Wait for response before returning to user-space
 	 * This can be optimized in future to even prepare response
-- 
2.17.1

